### PR TITLE
Restore hook for config change

### PR DESCRIPTION
**Pull request by anjumfatima90**

### CODE REVIEW

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 1]
```diff
@@ -227,6 +231,7 @@ public static WsLocationAdminImpl getInstance() {
      *                                      if bootstrap library location or instance root don't exist.
      */
     protected WsLocationAdminImpl(Map<String, Object> config) {
+        BundleContext bundleContext = FrameworkUtil.getBundle(getClass()).getBundleContext();
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: I would prefer this get passed in as an argument to the constructor.  Then for `com.ibm.ws.kernel.service.location.internal.WsLocationAdminImpl.createLocations(Map<String, Object>)` you would pass in `null`.   The code should handle when there is no BundleContext available.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 2]
```diff
@@ -335,7 +340,36 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
...
+                String newVariableSourceDirs = getEnv(BootstrapConstants.ENV_VARIABLE_SOURCE_DIRS);
+                if (!newVariableSourceDirs.equals(variableSourceDirs)) {
+                    SymbolRegistry.getRegistry().replaceStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, newVariableSourceDirs);
+                }
+            };
+        }, null);
+    }
+
+    String getEnv(final String key) {
+        if (key == null)
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: This null check is not necessary.  This method is only called with a non-null constant.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 3]
```diff
@@ -335,7 +340,36 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
...
+            value = AccessController.doPrivileged(new java.security.PrivilegedExceptionAction<String>() {
+                @Override
+                public String run() throws Exception {
+                    return System.getenv(key);
+                }
+            });
+        } catch (Exception ex) {
+
+        }
+        return value;
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: I don't see the need to use a PrivilegedExceptionAction here and with Java 8 you can reduce this to:
```java
        return AccessController.doPrivileged((PrivilegedAction<String>)() -> {
            return System.getenv(key);
        });
```

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java [hunk 4]
```diff
@@ -127,6 +130,17 @@
         WSConfigurationHelper wsConfigHelper = new WSConfigurationHelperImpl(metatypeRegistry, ce, bundleProcessor);
         registerService(bc, WSConfigurationHelper.class.getName(), wsConfigHelper);
 
+        // register restore hook to reprocess config if necessary
+        bc.registerService(CheckpointHookFactory.class, (p) -> new CheckpointHook() {
+            @Override
+            public void restore() {
+                try {
+                    reprocessConfig();
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java:138]: Will this notice that the variable dir setting changed?

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 5]
```diff
@@ -335,7 +340,36 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
 
         addResourcePath(bootstrapLib.getNormalizedPath());
 
-        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS));
+        String variableSourceDirs = (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS);
+        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, variableSourceDirs);
+
+        bundleContext.registerService(CheckpointHookFactory.class, (p) -> new CheckpointHook() {
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: Be sure to check for null bundle context.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java (hunk 2 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: done

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 6]
```diff
@@ -335,7 +339,30 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
 
         addResourcePath(bootstrapLib.getNormalizedPath());
 
-        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS));
+        String variableSourceDirs = (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS);
+        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, variableSourceDirs);
+
+        if (bundleContext == null) {
+            bundleContext = FrameworkUtil.getBundle(getClass()).getBundleContext();
+        }
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: Do not do this block of code.  If null is passed for bundleContext simply do not register the hook.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 7]
```diff
@@ -335,7 +339,30 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
 
         addResourcePath(bootstrapLib.getNormalizedPath());
 
-        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS));
+        String variableSourceDirs = (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS);
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: I would make this `final` to make it clear that it doesn't change.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java [hunk 8]
```diff
@@ -174,16 +188,17 @@ void start() throws ConfigUpdateException, ConfigValidationException, ConfigPars
             serverXMLConfig.loadInitialConfiguration(variableRegistry);
         }
 
-        final boolean reprocessConfig;
+        reprocessConfig();
+    }
+
+    void reprocessConfig() throws ConfigUpdateException {
         if (serverXMLConfig.isModified() || variableRegistry.variablesChanged()) {
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: I'm unsure if the call to `variableRegistry.variablesChanged()` will notice that the symbol for `WsLocationConstants.LOC_VARIABLE_SOURCE_DIR` has changed on restore.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java (hunk 3 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: done

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java (hunk 5 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: done

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java (hunk 6 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: done

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java (hunk 7 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: done

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java (hunk 8 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: I added a restore hook in ConfigVariableRegistry constructor to update the registry if variable_source_dir changes. `ConfigVariableRegistry.variablesChanged()` method compares the values in cache and registry. After my changes, the registry now has the updated values which will be compared to old cache entries.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java [hunk 9]
```diff
@@ -174,16 +188,17 @@ void start() throws ConfigUpdateException, ConfigValidationException, ConfigPars
...
             variableRegistry.clearVariableCache();
             changeHandler.updateAtStartup(serverXMLConfig.getConfiguration());
             serverXMLConfig.setConfigReadTime();
-            reprocessConfig = true;
-        } else {
-            reprocessConfig = false;
+            bundleProcessor.startProcessor(true);
         }
-        bundleProcessor.startProcessor(reprocessConfig);
+
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: This is the change I did, but I think it is wrong.

Before when the config bundle started we could call BundleProcessor.startProcessor(false), but now we will always use `true` or we will not call `startProcessor` at all.  I'm pretty sure that is not what we want here.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 10]
```diff
@@ -335,7 +339,26 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
...
-        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS));
+        final String variableSourceDirs = (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS);
+        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, variableSourceDirs);
+
+        if (bundleContext != null) {
+            bundleContext.registerService(CheckpointHookFactory.class, (p) -> new CheckpointHook() {
+                @Override
+                public void restore() {
+                    String newVariableSourceDirs = getEnv(BootstrapConstants.ENV_VARIABLE_SOURCE_DIRS);
+                    if (newVariableSourceDirs != null && !newVariableSourceDirs.equals(variableSourceDirs)) {
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: What about when `newVariableSourceDirs` is null.  I think there is a default location we would need to restore to.

Comment (anjumfatima90) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: I fixed the whole method. The way I was treating the environment variable earlier was wrong. There are multiple paths allowed in that variable.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java (hunk 9 above)
Comment (anjumfatima90) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: I passed a restoreHook flag to `SystemConfiguration.reprocessConfig(boolean)`

**Comment by anjumfatima90** [comment]
#build

**Comment by anjumfatima90** [comment]
#build

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 11]
```diff
@@ -335,7 +340,55 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
 
         addResourcePath(bootstrapLib.getNormalizedPath());
 
-        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS));
+        final String variableSourceDirs = (String) config.get(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS);
+        final String configDirStr = serverCfgDirStr;
+        SymbolRegistry.getRegistry().addStringSymbol(WsLocationConstants.LOC_VARIABLE_SOURCE_DIRS, variableSourceDirs);
+
+        if (bundleContext != null) {
+            //Service ranking of this hook here needs to less than service ranking of restore hook in "com.ibm.ws.config.xml.internal.SystemConfiguration.SystemConfiguration(BundleContext, SystemConfigSupport, ConfigurationAdmin). This is important in order to maintain the order of running the hooks"
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: Please split this code up on multiple, shorter lines.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java [hunk 12]
```diff
@@ -132,7 +135,8 @@
         registerService(bc, WSConfigurationHelper.class.getName(), wsConfigHelper);
 
         // register restore hook to reprocess config if necessary
-        bc.registerService(CheckpointHook.class, this, null);
+        // Service ranking of this hook here needs to than service ranking of restore hook in "com.ibm.ws.kernel.service.location.internal.WsLocationAdminImpl.WsLocationAdminImpl(BundleContext, Map<String, Object>)". This is important in order to maintain the order of running the hooks"
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: I'm one to complain about code comments, I make typos all the time in them :)

But I am unsure about this comment.  Does this ranking need to be great than or less than the one in WsLocationAdminImpl?  Also, please split this comment up into multiple lines.  I thought the code formatter would have done this for you.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java [hunk 13]
```diff
@@ -132,7 +135,8 @@
...
+        bc.registerService(CheckpointHook.class, this, FrameworkUtil.asDictionary(Collections.singletonMap(Constants.SERVICE_RANKING, 1000)));
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/SystemConfiguration.java]: This is not an issue your change introduced.  But this registration as well as the one above that registers the `WSConfigurationHelper` should both save the ServiceRegistration returned by registerService and then call ServiceRegistration.unregister on `SystemConfiguration.stop()`.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 14]
```diff
@@ -335,7 +340,55 @@ protected WsLocationAdminImpl(Map<String, Object> config) {
...
+            bundleContext.registerService(CheckpointHook.class, new CheckpointHook() {
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: This registration should be saved of for unregistration also.  Perhaps we should have WsLocationAdminImpl implement CheckpointHook and then register it as a CheckpointHook in com.ibm.ws.kernel.service.Activator.start.

File: dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java [hunk 15]
```diff
@@ -226,9 +234,9 @@ public static WsLocationAdminImpl getInstance() {
      * @throws IllegalStateException
      *                                      if bootstrap library location or instance root don't exist.
      */
-    protected WsLocationAdminImpl(Map<String, Object> config) {
+    protected WsLocationAdminImpl(BundleContext bundleContext, Map<String, Object> config) {
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: Where is the BundleContext used?

Comment (tjwatson) [reply dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/location/internal/WsLocationAdminImpl.java]: I pushed a commit to the branch to address this.

**Review by tjwatson — APPROVED** [review]

**Pull request merged**